Rename CPU /stats/summary metrics

ghodss opened the pull request:
In the new /stats/summary endpoint, rename UsageNanoCores and UsageCoreNanoSeconds to UsageRate and CumulativeNanoseconds respectively, ideally merging this into 1.2. Rationale:
- UsageNanoCores and UsageCoreNanoSeconds are significantly different in what they represent - one is nanoseconds sampled over time, the other is cumulative nanoseconds used since creation of the container - but the current names do not make this distinction clear at all.
- Clarify that UsageNanoCores is a rate by naming it UsageRate and that UsageCoreNanoSeconds is since the creation of the container by naming it CumulativeNanoseconds.
- Remove "Cores" since the fields are already namespace under "CPU".
- The API has not yet been released, so now is the ideal time to do a rename.

Apologies from the late bomb, and punting on this is not the end of the world, but it'd really be nice to get this backwards-incompatible rename out of the way.

I'm not attached to the new names so happy to take further suggestions.

@dchen1107 @timstclair @vishh @bgrant0607

Commit:
Rename CPU /stats/summary metrics

derekwaynecarr commented:
/cc @derekwaynecarr @kubernetes/rh-cluster-infra in case this gets in 1.2 so we can track.

timstclair commented on pkg/kubelet/api/v1alpha1/stats/types.go:
```diff
@@ -119,9 +119,9 @@ type CPUStats struct {
+	UsageRate *uint64 `json:"usageNanoCores,omitempty"`
 	// Cumulative CPU usage (sum of all cores) since object creation.
-	UsageCoreNanoSeconds *uint64 `json:"usageCoreNanoSeconds,omitempty"`
+	CumulativeNanoseconds *uint64 `json:"usageCoreNanoSeconds,omitempty"`
```
I'd like to keep cores in the units here to make it clear what we're measuring (`UsageRateCores` and `CumulativeCoreNanoseconds`)

timstclair replied:
We're measuring core-seconds, not cpu-seconds, which matters on multicore processors. I suppose that's not entirely accurate with hyperthreading. Do you think this is a given and redundant? I also like including cores because it makes it clear that `UsageRate` is *not* a unitless quantity.

dchen1107 replied:
I would like including core in the name too. That is the intention what we want to measure even it is not accurate with hyperthreading today.

ghodss replied:
@vishh Oversight, updating them now, along with the e2e tests.

So are we settled on `UsageRateCores` and `CumulativeCoreNanoseconds`?

vishh replied:
+1. We need to update heapster as well and make a new release @timstclair.

ghodss replied:
Or perhaps `CoreUsageRate` and `CumulativeCoreNanoseconds`? Reads a little better.

dchen1107 commented:
@ghodss I am fine with the new names (UsageRateCores and CumulativeCoreNanoseconds) now, but open for other's input. 

But we cannot simply change those names in kubelet like this way without updating heapster; otherwise, e2e will failed. Here is my suggestion: 

* Once we settle down with the new names, you have to reconfigure heapster with --source=kubernetes:'' so that heapster will temporarily switch back to old endpoint. 
* Then we can update heapster to understand the new names and bump its version in kubernetes cluster with --source=kubernetes.summary_api:''

fgrzadkowski commented:
Is this renaming really necessary? This will break our e2e tests, require very last minute fixes etc. Also if this unofficial API maybe we can just use different names in the final kubelet metric api?

Commit:
Revert "Use the summary_api with heapster by default"

This reverts commit 1d90394bf67e4da6948e30b53e830d0d930f60f0.

ghodss commented on cluster/addons/cluster-monitoring/google/heapster-controller.yaml:
```diff
@@ -38,7 +38,7 @@ spec:
               memory: {{ heapster_memory }}
           command:
             - /heapster
-            - --source=kubernetes.summary_api:''
```
This is temporary until we have a chance to update and release heapster.

ghodss commented:
* The two fields have been renamed to CoreUsageRate and CumulativeCoreNanoseconds.
* Fixed JSON tags.
* Reverted e2e heapster config that points heapster at summary API. This will have to get reverted back after this PR is merged and hipster is updated to use the new fields.

I agree this is a lot of churn last minute and am happy to abort, but am also happy to do whatever work is necessary to push this through, again since it's an unreleased API and it's easiest to make this change now.

dchen1107 commented:
cc/ @erictune Your suggestion on naming here. Thanks!

erictune commented:
I agree that having both a cumulative and rate measurements is going to be confusing for some people.  

However, "cpu  seconds" implies cumulative to me.  If I google for `cpu seconds` I get a one-box that summarizes the wikipedia page on [cpu time](https://en.wikipedia.org/wiki/CPU_time).  I think `core nanoseconds" also implies cumulative by extension.  So, I'm not sold on the premise of this change.

Since we are discussing these names, though, I don't like the current names, for other reasons.

I'd rather say "cpu" than "core".  Google says that `"cpu seconds"` (~86,000 hits) is much more common than `core seconds`  (~2400 hits).   I'd tend for the more common term.  Also, the request and limit are called `cpu`, not `core`, so it would be better to be consistent there.  Also, on hyperthreaded (SMT) processors, core means something different than what I think this is measuring.  Based on that, remove the word cores and either replace with cpu, or drop it altogether since the parent contains it.  

I don't care much if those changes happens in this PR, or later when promoting this API to v1beta1.

I'm not sure that the word "usage" is helpful here.  It isn't distinguishing from anything else.  (For memory, `usage` is opposed to `workingset`.  For networking, we don't call it usage, but it is not a request or a limit.)

ghodss commented:
After staring at this all day, I realize why I didn't understand the naming originally - the idea is Usage[of]&lt;Unit&gt;, so it's Usage of NanoCores (which is Core Nanoseconds per Second) over a sample window, and Usage of CoreNanoSeconds which is total time spent on the CPU. It's still not totally intuitive to me so I've made some tweaks to the comments in #22832 to help.

Now that's it clearer to me, plus all the different opinions about the names, plus the overhead of coordinating with heapster, plus the imminent 1.2 deadline, I think we can close this PR for now and worry about this during v1beta1 promotion (as per @erictune). We'll also probably need a new PR since we'd likely want to namespace /stats/summary into /api/v1beta1/stats/summary at that point as well, since I don't think it makes sense to actually break this API after release.

We can reopen this and make the change now if anyone strongly disagrees.

timstclair commented:
> I agree that having both a cumulative and rate measurements is going to be confusing for some people.

I'd consider simply removing the rate metric as it's (1) not used, and (2) poorly defined (depends on the sample rate).

@erictune Thanks for explaining your reasoning behind the naming, it will be helpful for me when thinking about future names, even if it's too late to fix these ones.

> ... we'd likely want to namespace /stats/summary into /api/v1beta1/stats/summary at that point ...

Yes, I'd like to add versioned endpoints for the summary and other cAdvisor APIs. Look for a formal proposal from me sometime next week...

@ghodss Thanks for letting this one go for now... naming is hard!
